src/ConstructedStyleSheet.ts

```diff
@@ -0,0 +1,240 @@
+  sheet: ConstructedStyleSheet,
+  location: Location,
+): HTMLStyleElement {
+  var adopter = document.createElement('style');
+  $adoptersByLocation.get(sheet)!.set(location, adopter);
+  $locations.get(sheet)!.push(location);
+
+  return adopter;
+}
+
+export function getAdopterByLocation(
+  sheet: ConstructedStyleSheet,
+  location: Location,
+): HTMLStyleElement | undefined {
+  return $adoptersByLocation.get(sheet)!.get(location);
+}
+
+export function removeAdopterLocation(
+  sheet: ConstructedStyleSheet,
+  location: Location,
+): void {
+  $adoptersByLocation.get(sheet)!.delete(location);
+  $locations.set(
+    sheet,
+    $locations.get(sheet)!.filter(function (_location) {
+      return _location !== location;
+    }),
+  );
+}
+
+/**
+ * Re-styles a single `<style>` adopter according to the basic style sheet.
+ *
+ * NOTE: don't use it for disconnected adopters. It will throw an error.
+ */
+export function restyleAdopter(
+  sheet: ConstructedStyleSheet,
+  adopter: HTMLStyleElement,
+): void {
+  clearRules(adopter.sheet!);
+  /*#__INLINE__*/ insertAllRules($basicStyleSheet.get(sheet)!, adopter.sheet!);
+}
+
+/*
+ * Private methods
+ */
+
+/**
+ * This method checks if the method of the class is called correctly using the
+ * CSSStyleSheet instance. It is necessary to be sure that all private
+ * properties are initialized.
+ */
+function checkInvocationCorrectness(self: ConstructedStyleSheet) {
+  if (!$basicStyleSheet.has(self)) {
+    throw new TypeError('Illegal invocation');
+  }
+}
+
+/**
+ * A replacement for CSSStyleSheet class which is not actually constructable in
+ * any browser that does not support Constructable Style Sheet proposal. To
+ * fulfil the Constructable Style Sheet specification, it preserves a separate
+ * HTMLStyleElement that behaves as a backup for all the `<style>` adopters.
+ */
+declare class ConstructedStyleSheet extends CSSStyleSheet {
+  replace(text: string): Promise<ConstructedStyleSheet>;
+  replaceSync(text: string): void;
+}
+
+function ConstructedStyleSheet(this: ConstructedStyleSheet) {
+  var style = document.createElement('style');
+  bootstrapper.body.appendChild(style);
+
+  // Init private properties
+  $basicStyleSheet.set(this, style.sheet!);
+  $locations.set(this, []);
+  $adoptersByLocation.set(this, new WeakMap());
+}
+
+var proto = ConstructedStyleSheet.prototype;
+
+proto.replace = function replace(
+  contents: string,
+): Promise<ConstructedStyleSheet> {
+  try {
+    this.replaceSync(contents);
+
+    return Promise.resolve(this);
+  } catch (e) {
+    return Promise.reject(e);
+  }
+};
+
+proto.replaceSync = function replaceSync(contents) {
+  // CSSStyleSheet.prototype.replaceSync('body {}')
+  checkInvocationCorrectness(this);
+
+  if (typeof contents === 'string') {
+    var self = this;
+
+    var basic = $basicStyleSheet.get(self)!;
+    var sanitized = rejectImports(contents);
+    clearRules(basic);
+
+    if (sanitized) {
+      basic.insertRule(sanitized, 0);
```

Comment:
Unless I'm missing something this will fail if there is more than one rule in the stylesheet. Changing the test to 

```javascript
headingStyles
  .replace(
    ` h1 {
    color: tomato;
}  body { background: mediumaquamarine; }`,
  )
  .then(sheet => {
    console.log(`${sheet} styles have been replaced.`);
    console.assert(sheet === headingStyles, 'not the same', 'yass');
  });
```

will throw, hence the original `innerHTML` call. We could potentially parse the thing with a regexp and see if insertRule is faster, but I doubt it.

Reply:
Good catch, thanks! I replaced it with the `basic.ownerNode.textContent = sanitized`. It will do all the parsing job for us and still allow using `restyleAdopter` with `insertRule` approach.